gravitee-am-gateway/gravitee-am-gateway-handler/gravitee-am-gateway-handler-core/src/main/java/io/gravitee/am/gateway/handler/root/resources/handler/login/LoginSelectionRuleHandler.java

```diff
@@ -0,0 +1,110 @@
+import static io.gravitee.am.gateway.handler.root.resources.handler.login.LoginSocialAuthenticationHandler.SOCIAL_PROVIDER_CONTEXT_KEY;
+
+/**
+ * @author Eric LELEU (eric.leleu at graviteesource.com)
+ * @author GraviteeSource Team
+ */
+public class LoginSelectionRuleHandler  implements Handler<RoutingContext>  {
+    private final Logger logger = LoggerFactory.getLogger(this.getClass());
+
+    private final boolean fromIdentifierFirstLogin;
+
+    public LoginSelectionRuleHandler(boolean fromIdentifierFirstLogin) {
+        this.fromIdentifierFirstLogin = fromIdentifierFirstLogin;
+    }
+
+    @Override
+    public void handle(RoutingContext routingContext) {
+        final Client client = routingContext.get(ConstantKeys.CLIENT_CONTEXT_KEY);
+
+        List<IdentityProvider> socialProviders = routingContext.get(SOCIAL_PROVIDER_CONTEXT_KEY);
+        if ((socialProviders != null && !socialProviders.isEmpty())
+                && (client.getIdentityProviders() != null && !client.getIdentityProviders().isEmpty())) {
+
+            var appIdpMap = client.getIdentityProviders().stream().collect(Collectors.toMap(
+                    ApplicationIdentityProvider::getIdentity, Function.identity()
+            ));
+
+            var context = new SimpleAuthenticationContext(new VertxHttpServerRequest(routingContext.request().getDelegate()), routingContext.data());
+            var templateEngine = context.getTemplateEngine();
+            var identityProvider = socialProviders.stream()
+                    .filter(idp -> appIdpMap.containsKey(idp.getId()))
+                    .filter(idp -> evaluateRule(appIdpMap.get(idp.getId()), templateEngine, idp))
+                    .findFirst();
+
+            if (identityProvider.isPresent()) {
+                Map<String, String> urls = routingContext.get(SOCIAL_AUTHORIZE_URL_CONTEXT_KEY);
+                UriBuilder uriBuilder = UriBuilder.fromHttpUrl(urls.get(identityProvider.get().getId()));
+
+                if (fromIdentifierFirstLogin) {
+                    // encode login_hint parameter for external provider (Azure AD replace the '+' sign by a space ' ')
+                    uriBuilder.addParameter(Parameters.LOGIN_HINT, UriBuilder.encodeURIComponent(routingContext.request().getParam(USERNAME_PARAM_KEY)));
+                }
+
+                doRedirect(routingContext, uriBuilder.buildString());
+                return;
+            }
+        }
+
+        routingContext.next();
+    }
+
+    private void doRedirect(RoutingContext routingContext, String url) {
+        routingContext.response()
+                .putHeader(io.vertx.core.http.HttpHeaders.LOCATION, url)
+                .setStatusCode(302)
+                .end();
+    }
+
+    private boolean evaluateRule(ApplicationIdentityProvider appIdp, io.gravitee.el.TemplateEngine templateEngine, IdentityProvider idp) {
```

Comment:
The `IdentityProvider idp` parameter is useless
```suggestion
    private boolean evaluateRule(ApplicationIdentityProvider appIdp, io.gravitee.el.TemplateEngine templateEngine) {
```